src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,702 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ *  CEPH messenger engine
+ *
+ *  FIO engine which uses ceph messenger as a transport.  See corresponding
+ *  FIO client and server jobs for details.
+ */
+
+#include "global/global_init.h"
+#include "msg/Messenger.h"
+#include "messages/MOSDOp.h"
+#include "messages/MOSDOpReply.h"
+#include "common/perf_counters.h"
+#include "ring_buffer.h"
+
+#include <fio.h>
+#include <flist.h>
+#include <optgroup.h>
+
+enum ceph_msgr_type {
+  CEPH_MSGR_TYPE_UNDEF,
+  CEPH_MSGR_TYPE_POSIX,
+  CEPH_MSGR_TYPE_DPDK,
+  CEPH_MSGR_TYPE_RDMA,
+};
+
+const char *ceph_msgr_types[] = { "undef", "async+posix",
+				  "async+dpdk", "async+rdma" };
+
+struct ceph_msgr_options {
+  struct thread_data *td__;
+  unsigned int is_receiver;
+  unsigned int is_single;
+  unsigned int port;
+  const char *hostname;
+  const char *conffile;
+  enum ceph_msgr_type ms_type;
+};
+
+class FioDispatcher;
+
+struct ceph_msgr_data {
+  ceph_msgr_data(struct ceph_msgr_options *o_, unsigned iodepth) :
+    o(o_), msgr(NULL), disp(NULL),
+    io_inflight_nr(0),
+    io_pending_nr(0) {
+    INIT_FLIST_HEAD(&io_inflight_list);
+    INIT_FLIST_HEAD(&io_pending_list);
+    ring_buffer_init(&io_completed_q, iodepth);
+    pthread_spin_init(&spin, PTHREAD_PROCESS_PRIVATE);
+  }
+
+  struct ceph_msgr_options *o;
+  Messenger *msgr;
+  FioDispatcher *disp;
+  pthread_spinlock_t spin;
+  struct ring_buffer io_completed_q;
+  struct flist_head io_inflight_list;
+  struct flist_head io_pending_list;
+  unsigned int io_inflight_nr;
+  unsigned int io_pending_nr;
+};
+
+struct ceph_msgr_io {
+  struct flist_head list;
+  struct ceph_msgr_data *data;
+  struct io_u *io_u;
+  MOSDOp *req_msg; /** Cached request, valid only for sender */
+};
+
+struct ceph_msgr_reply_io {
+  struct flist_head list;
+  MOSDOpReply *rep;
+};
+
+typedef union {
+  uint64_t uintptr;
+  void     *ptr;
+} ptr_t;
+
+static void *hex_to_ptr(const std::string &hex)
+{
+  ptr_t p;
+
+  assert(hex.length() == sizeof(p) * 2);
+  sscanf(hex.c_str(), "%" SCNx64, &p.uintptr);
```

Comment:
formatting I/O is known to be quite ineffective AFAIK. And this function seems to be on the main path. May be replace with strtol? Or even use it directly.

Reply:
This is weird indeed, but objid is a std::string, and it won't work if I pass binary (obviously with possible 0x0 chars) in it. So directly won't work.  But of course I can give a try with strtol, but frankly I do not see scanf in perf tops. So do not care a lot.

Reply:
BTW you were right, strtoul() gives ~10mb/s and 0.1ms drop in latency. Good hint, thanks.